docs/cardano-monads.rst

```diff
@@ -0,0 +1,772 @@
+it's done at use site), and better predictability.
+
+Verdict: preferable to ``ReaderT`` as long as we can put up with reduced
+convenience.
+
+ReaderT over Abstract Base
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After a discussion, @georgeee and @int-index came up with a hybrid
+approach that combines many elements of other approaches to provide a
+comprehensive solution.
+
+Design goals (in addition to already discussed criteria):
+
+* avoid data types parametrized by an abstract monad ``m``
+  (including explicit method dictionaries),
+  because it makes reasoning about code complicated
+
+* avoid hard-coded ``IO``, ``MonadIO``, ``MonadBaseControl IO``, etc,
+  in order to have good mocking capabilities and separation of concerns
+
+And here are some observations:
+
+* operationally, our monads can be ``ReaderT ctx IO`` for production code
+  and ``ReaderT ctx (ExceptT SomeException ST)`` (or similar) in
+  pure tests. Both of these provide (1) passing context, (2) throwing/catching
+  exceptions, (3) mutable state.
+
+* runtime configuration of effects arises for CLI parameters or YAML configuration
+  (for instance, the slotting method is affected by the '--no-ntp' flag). For this
+  use cases effect context sums must be sufficient.
+
+* presumably, any effect dictionary can be separated into its closure (context)
+  and top-level operations
+
+With these facts established, let us consider the approach from multiple angles:
+how effect definitions would look like, and how effect implementatinos look like.
+
+First, let us have two base monads::
+
+    newtype Prod a = Prod (IO a)
+    newtype Test a = Test (ExceptT SomeException ST a)
+
+To define an effect, in the most general case we will need two classes and a type family.
+In the first class we define the methods, mtl-style. The second class is a ``Has``-class to
+extract a part of a context. And a type family determines what context we need for the effect::
+
+    class MonadFoo b where
+        type FooContext b
+        sepulate :: HasFooContext b ctx => ReaderT ctx b Sepulki
+        patanize :: HasFooContext b ctx => ctx
+                                        -> ReaderT ctx b a
+                                        -> ReaderT ctx b (Either Patak a)
+
+    class HasFooContext b ctx where
+        fooContext :: ctx -> FooContext b
+
+A few things to notice:
+
+* the parameter ``b`` stands for "base monad", it will be instantiated to
+  something simple like ``Prod`` or ``Test``.
+
+* all methods are defined in the ``ReaderT ctx b`` monad.
+
+* we are fine with this monad being used in a negative position (in
+  ``patanize``), because it is not meant to be lifted/unlifted in any
+  circumstances.
+
+This is how we could define database-related monads this way::
+
+    class MonadDBGet b where
+        type DBContext b
+        dbGet :: HasDBContext b ctx => DBTag -> ByteString -> ReaderT ctx (Base b) (Maybe ByteString)
```

Comment:
What is `Base`?

Reply:
It was a leftover from a previous version, thanks, removed.